File: src/main/groovy/cash/andrew/mntrailconditions/data/DataModule.java
package cash.andrew.mntrailconditions.data;

import android.app.Application;
import android.content.SharedPreferences;
import cash.andrew.mntrailconditions.data.moshi.adapters.LocalDateTimeJsonAdapter;
import cash.andrew.mntrailconditions.data.okhttp.ApiVersionHeaderInterceptor;
import cash.andrew.mntrailconditions.data.okhttp.UserAgentInterceptor;
import com.f2prateek.rx.preferences.RxSharedPreferences;
import com.jakewharton.picasso.OkHttp3Downloader;
import cash.andrew.mntrailconditions.data.api.ApiModule;
import com.squareup.moshi.Moshi;
import com.squareup.picasso.Picasso;
import dagger.Module;
import dagger.Provides;
import java.io.File;
import java.util.concurrent.TimeUnit;
import javax.inject.Singleton;
import okhttp3.Cache;
import okhttp3.OkHttpClient;
import timber.log.Timber;

import static android.content.Context.MODE_PRIVATE;
import static com.jakewharton.byteunits.DecimalByteUnit.MEGABYTES;

@Module(
    includes = ApiModule.class,
    complete = false,
    library = true
)
public final class DataModule {
  static final int DISK_CACHE_SIZE = (int) MEGABYTES.toBytes(50);

  @Provides @Singleton SharedPreferences provideSharedPreferences(Application app) {
    return app.getSharedPreferences("cash.andrew.mntrailconditions", MODE_PRIVATE);
  }

  @Provides @Singleton RxSharedPreferences provideRxSharedPreferences(SharedPreferences prefs) {
    return RxSharedPreferences.create(prefs);
  }

  @Provides @Singleton Moshi provideMoshi() {
    return new Moshi.Builder()
        .add(new LocalDateTimeJsonAdapter())
        .build();
  }

  @Provides @Singleton OkHttpClient provideOkHttpClient(Application app) {
    return createOkHttpClient(app).build();
  }

  @Provides @Singleton Picasso providePicasso(Application app, OkHttpClient client) {
    return new Picasso.Builder(app)
        .downloader(new OkHttp3Downloader(client))
        .listener((picasso, uri, e) -> Timber.e(e, "Failed to load image: %s", uri))
        .build();
  }

  static OkHttpClient.Builder createOkHttpClient(Application app) {
    // Install an HTTP cache in the application cache directory.
    File cacheDir = new File(app.getCacheDir(), "http");
    Cache cache = new Cache(cacheDir, DISK_CACHE_SIZE);

    return new OkHttpClient.Builder()
        .cache(cache)
        .addInterceptor(new ApiVersionHeaderInterceptor())
        .addInterceptor(new UserAgentInterceptor());
  }
}

File: src/main/kotlin/cash/andrew/mntrailconditions/data/DataModule.kt
package cash.andrew.mntrailconditions.data

import android.app.Application
import android.content.Context.MODE_PRIVATE
import android.content.SharedPreferences
import cash.andrew.mntrailconditions.data.api.ApiModule
import cash.andrew.mntrailconditions.data.moshi.adapters.InstantJsonAdapter
import cash.andrew.mntrailconditions.data.moshi.adapters.LocalDateTimeJsonAdapter
import cash.andrew.mntrailconditions.data.okhttp.UserAgentInterceptor
import com.f2prateek.rx.preferences2.Preference
import com.f2prateek.rx.preferences2.RxSharedPreferences
import com.jakewharton.byteunits.DecimalByteUnit.MEGABYTES
import com.readystatesoftware.chuck.ChuckInterceptor
import com.squareup.moshi.Moshi
import dagger.Module
import dagger.Provides
import okhttp3.Cache
import okhttp3.OkHttpClient
import java.io.File
import javax.inject.Qualifier
import javax.inject.Singleton

@Module(includes = [ApiModule::class])
object DataModule {

    private val DISK_CACHE_SIZE = MEGABYTES.toBytes(50).toInt()

    @JvmStatic
    @Provides
    @Singleton
    fun provideSharedPreferences(app: Application): SharedPreferences =
            app.getSharedPreferences("cash.andrew.mntrailconditions", MODE_PRIVATE)

    @JvmStatic
    @Provides
    @Singleton
    fun provideRxSharedPreferences(prefs: SharedPreferences) = RxSharedPreferences.create(prefs)

    @JvmStatic
    @Provides
    @Singleton
    fun provideMoshiBuilder(): Moshi.Builder = Moshi.Builder()
                .add(InstantJsonAdapter())
                .add(LocalDateTimeJsonAdapter())

    @JvmStatic
    @Provides
    @Singleton
    fun provideOkHttpClient(app: Application): OkHttpClient = createOkHttpClient(app).build()

    @JvmStatic
    @SavedTrails
    @Provides
    @Singleton
    fun provideTrailFavorites(prefs: RxSharedPreferences): Preference<Set<String>> =
            prefs.getStringSet("trail-favorites")

    @JvmStatic
    @Provides
    @NotificationTrails
    @Singleton
    fun provideNotificationList(prefs: RxSharedPreferences): Preference<Set<String>> =
            prefs.getStringSet("trail-notifications")

    private fun createOkHttpClient(app: Application): OkHttpClient.Builder =
            OkHttpClient.Builder()
                .cache(Cache(File(app.cacheDir, "http"), DISK_CACHE_SIZE.toLong()))
                .addInterceptor(UserAgentInterceptor())
                .addInterceptor(ChuckInterceptor(app).showNotification(false))
}

@Qualifier annotation class SavedTrails
@Qualifier annotation class NotificationTrails
